File: librefit-web/src/lib/api/wizard.ts
import { getDateAsStr, parseStringAsDate } from '$lib/date';
import { CalculationGoal } from '$lib/model';
import type {
	CalorieTarget,
	NewCalorieTarget,
	NewWeightTarget,
	WeightTarget,
	Wizard,
	WizardInput,
	WizardResult, WizardTargetDateInput,
	WizardTargetDateResult, WizardTargetWeightInput,
	WizardTargetWeightResult
} from '$lib/model';
import { isAfter } from 'date-fns';
import { invoke } from '@tauri-apps/api/core';

export const calculateTdee = async (wizardInput: WizardInput): Promise<WizardResult> => {
	wizardInput.age = +wizardInput.age;
	wizardInput.weight = +wizardInput.weight;
	wizardInput.height = +wizardInput.height;
	wizardInput.weeklyDifference = +wizardInput.weeklyDifference;

	return invoke('wizard_calculate_tdee', {
		input: wizardInput
	});
};

/**
 * @param wizard {Wizard}
 * @returns {Promise<void>}
 */
export const postWizardResult = async (wizard: Wizard): Promise<void> => {
	return invoke('wizard_create_targets', {
		input: wizard
	});
};

export const calculateForTargetDate = async (wizardTargetDateInput: WizardTargetDateInput): Promise<WizardTargetDateResult> => {
	return invoke('wizard_calculate_for_target_date', {
		input: wizardTargetDateInput
	});
};

export const calculateForTargetWeight = async (wizardTargetWeightInput: WizardTargetWeightInput): Promise<WizardTargetWeightResult> => {
  wizardTargetWeightInput.targetWeight = +wizardTargetWeightInput.targetWeight;
  wizardTargetWeightInput.startDate = "2024-12-31";

	return invoke('wizard_calculate_for_target_weight', {
		input: wizardTargetWeightnput
	});
};

/**
 * @param wizardInput {WizardInput}
 * @param wizardResult {WizardResult}
 * @param customWizardResult {WizardTargetWeightResult}
 * @param startDate {Date}
 * @param targetWeight {Number}
 * @returns Map
 */
export const createTargetWeightTargets = (
	wizardInput: WizardInput,
	wizardResult: WizardResult,
	customWizardResult: WizardTargetWeightResult,
	startDate: Date,
	targetWeight: number
): Map<string, { calorieTarget: CalorieTarget; weightTarget: WeightTarget }> => {
	const rates = Object.keys(customWizardResult.resultByRate);
	const multiplier = targetWeight < wizardInput.weight ? -1 : 1;
	const todayStr = getDateAsStr(new Date());

	return rates.reduce((acc, rate) => {
		const calorieTarget: NewCalorieTarget = {
			added: todayStr,
			startDate: getDateAsStr(startDate),
			endDate: customWizardResult.dateByRate[rate],
			targetCalories: wizardResult.tdee + multiplier * parseFloat(rate),
			maximumCalories: wizardResult.tdee
		};

		const weightTarget: NewWeightTarget = {
			added: todayStr,
			startDate: getDateAsStr(startDate),
			endDate: customWizardResult.dateByRate[rate],
			initialWeight: wizardInput.weight,
			targetWeight: targetWeight
		};

		acc.set(rate, { calorieTarget, weightTarget });

		return acc;
	}, new Map());
};

export const createTargetDateTargets = (
	wizardInput: WizardInput,
	wizardResult: WizardResult,
	customWizardResult: WizardTargetDateResult,
	startDate: Date,
	endDateStr: string
): Map<string, { calorieTarget: CalorieTarget; weightTarget: WeightTarget }> => {
  console.log(customWizardResult);

	const rates = Object.keys(customWizardResult.resultByRate);
	const multiplier = wizardInput.calculationGoal === CalculationGoal.Loss ? -1 : 1;
	const todayStr = getDateAsStr(new Date());

	return rates.reduce((acc, rate) => {
		const rateWizardResult = customWizardResult.resultByRate[rate] as WizardResult;

		const calorieTarget: NewCalorieTarget = {
			added: todayStr,
			startDate: getDateAsStr(startDate),
			endDate: endDateStr,
			targetCalories: wizardResult.tdee + multiplier * parseFloat(rate),
			maximumCalories: wizardResult.tdee
		};

		const weightTarget: NewWeightTarget = {
			added: todayStr,
			startDate: getDateAsStr(startDate),
			endDate: endDateStr,
			initialWeight: wizardInput.weight,
			targetWeight: rateWizardResult.targetWeight
		};

    acc.set(rate, { calorieTarget, weightTarget });

		return acc;
	}, new Map());
};

export const validateCustomWeight = (detail: { value: number }): { valid: boolean; errorMessage?: string } => {
	if (detail.value < 30 || detail.value > 300) {
		return {
			valid: false,
			errorMessage: 'Please provide a weight between 30kg and 300kg.'
		};
	}

	return { valid: true };
};

export const validateCustomDate = (detail: { value: string }): { valid: boolean; errorMessage?: string } => {
	if (isAfter(new Date(), parseStringAsDate(detail.value))) {
		return {
			valid: false,
			errorMessage: 'Your target date lies in the past.'
		};
	}

	return { valid: true };
};

File: src-tauri/src/crud/cmd/wizard.rs
use crate::calc;
use crate::crud;
use crate::crud::cmd::error_handler::handle_error;
use calc::wizard::{
    Wizard, WizardInput, WizardResult, WizardTargetDateInput, WizardTargetDateResult,
    WizardTargetWeightInput, WizardTargetWeightResult,
};
use crud::db;
use crud::db::connection::create_db_connection;
use diesel::Connection;
use tauri::command;
use validator::ValidationErrors;

#[command]
pub fn wizard_calculate_tdee(input: WizardInput) -> Result<WizardResult, ValidationErrors> {
    log::info!(">>> wizard_calculate_tdee: {:?}", input);

    let wizard_result = calc::wizard::calculate(input);

    log::info!(">>> result={:?}", wizard_result);

    wizard_result
}

#[command]
pub fn wizard_create_targets(input: Wizard) -> Result<(), String> {
    log::info!(">>> wizard_create_targets: {:#?}", input);

    let manager = &mut create_db_connection();

    manager
        .transaction(|conn| {
            db::repo::weight::create_weight_target(conn, &input.weight_target)?;
            db::repo::weight::create_weight_tracker_entry(conn, &input.weight_tracker)?;
            db::repo::calories::create_calorie_target(conn, &input.calorie_target)?;

            Ok(())
        })
        .map_err(handle_error)
}

#[command]
pub fn wizard_calculate_for_target_date(
    input: WizardTargetDateInput,
) -> Result<WizardTargetDateResult, ValidationErrors> {
    log::info!(">>> wizard_calculate_for_target_date: {:?}", input);

    let wizard_result = calc::wizard::calculate_for_target_date(&input);

    log::info!(">>> result={:?}", wizard_result);

    return wizard_result;
}

#[command]
pub fn wizard_calculate_for_target_weight(
    input: WizardTargetWeightInput,
) -> Result<WizardTargetWeightResult, ValidationErrors> {
    log::info!(">>> wizard_calculate_for_target_weight {:?}", input);

    let wizard_result = calc::wizard::calculate_for_target_weight(&input);

    log::info!(">>> result={:?}", wizard_result);

    return wizard_result;
}
